### PR TITLE
Overloading of comparison operator for pauli operator class. 

### PR DESCRIPTION
**Pull request by afu1**
This allows for direct comparison of pauli operators and is much faster than conversion to text as used previously. 

## Description
in qiskit.tools.qi.pauli I added the two overloading operators to the Pauli class for == and !=
in qiskit.tools.apps.fermion I changed the textual comparison that was used to a direct one

## Motivation and Context

Some code in the chemistry examples uses a comparison 'if pauli1.to_label() == pauli2.to_label().'
With the proposed change the comparison is just 'if pauli1 == pauli2'. The direct comparison is 15x faster than the conversion to strings and thus significantly speeds up several codes.

## How Has This Been Tested?
pa = Pauli([0,0,1,1,1],[1,0,1,0,0])
pb = Pauli([0,0,1,1,1],[1,0,1,0,0])
pc = Pauli([0,0,1,1,1],[1,0,1,0,1])
if pa == pb:
    print('the first two are equal')
if pa != pc:
    print('the second two are not equal')

and I tested it on a new version of QChem Notebook which gave a 15x improvement when generating the Paulis for larger molecules. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by afu1** [comment]
This change seems to break the function group_paulis in qiskit.tools.apps.optimization.

**Comment by diego-plan9** [comment]
Thanks, @afu1 !

> This change seems to break the function group_paulis in qiskit.tools.apps.optimization.

Could you provide more information about the breaking (ideally, including a code sample in your reply)?

File: qiskit/tools/qi/pauli.py [hunk 1]
```diff
@@ -60,6 +60,23 @@ def __str__(self):
...
 
+    def __eq__(self, other):
+        """Return True if all Pauli terms are equal."""
+        bres = False
+        if self.numberofqubits == other.numberofqubits:
+            if self.v == other.v and self.w == other.w:
+                    bres = True
+        return bres
+
+    def __ne__(self, other):
```
Comment (diego-plan9) [inline qiskit/tools/qi/pauli.py]: Is defining `__ne__` really needed? By default, Python 3 will invert the result of `__eq__` - could you check and remove `__ne__` if that is the case?

**Comment by nikomoll** [comment]
It doesn't break anything in my case adding the comparison. To compare the arrays, on should use the function np.all(). The code would look like this:

    def __eq__(self, other):
        """Return True if all Pauli terms are equal."""
        bres = False
        if self.numberofqubits == other.numberofqubits:
            if np.all(self.v == other.v) and np.all(self.w == other.w):
                    bres = True
        return bres

**Comment by afu1** [comment]
I had an incompatible version of qiskit.tools.apps.optimization in my environment. I agree with @nikomoll that the change is compatible with the correct unmodified version.  I removed the overloading for != as suggested since it does not seem to be needed. I also included the suggested change with np.all() and checked that it works.

**Comment by diego-plan9** [comment]
Thanks for the quick turnaround (and @nikomoll  for the note about `np.all()`) - it's looking good!

Could you also add some tests:
* one or two for the equality comparison (something similar to the excerpt in the *"How Has This Been Tested?"* section on the PR description)
* one that executes `group_paulis()`

to the existing `test.python.test_apps.TestPauli` test case? It seems we are having a small surge in Pauli-related activity on the repository, and it would be awesome if we could take the chance to make the tests a bit more formal.

**Comment by afu1** [comment]
@diego-plan9 the testcase test.python.test_apps.TestPauli does not exist. I can put a test case in test.python.test_qi.TestPauli for the comparison. If you actually mean group_paulis() and not pauli_group() then that would go in test.python.test_apps.TestQuantumOptimization .

**Comment by diego-plan9** [comment]
Sorry for the confusion, I intended to:
* point to the existing `test.python.test_qi.TestPauli` case for the comparison test.
* for the other test, I mentioned `group_paulis()` based on your earlier comment (even if it was later dismissed as an unrelated issue - just for safety) - and you are right that it would  be more consistent to add it to  `test.python.test_apps.TestQuantumOptimization`.

Thanks for catching it early!

**Comment by jaygambetta** [comment]
I agree with @antoniomezzacapo but can we add tests for the comparison and multiplication before we merge.

**Comment by antoniomezzacapo** [comment]
I have added tests here https://github.com/QISKit/qiskit-sdk-py/pull/166

**Comment by diego-plan9** [comment]
Thanks everyone for the reviews and the contributions!

For making managing the PRs a bit easier and preserving contributions, I  have included the tests on this PR, and rebased it against the current master - in practice, @afu1 , it has the side effect that your `master` branch has been force-push and modified. For collaborative pull requests such as this one where several people might be working on the same, it might be handy for the people involved to work on the same branch directly - please let us know if it would make sense or if any help is needed in that regard.

So everything is looking good, has been reviewed and approved, tests have been included, and working, so ... merging!

**Pull request merged**